Add FlatBufferBuilder move semantics tests to the main test suite

sutambe opened the pull request:
This is a follow up PR to #4893. Moving `FlatBufferBuilder` move constructor and move assignment tests to the main test suite (so that they are run with every change). `MessageBuilder` tests remain in `./grpc/tests`. Refactored the main test suite a bit.

sutambe commented:
@gabyx @aardappel  I've an observation/suggestion regarding `release` and `release_raw` functions in `vector_downward`. I think both release functions are a little too aggressive regarding setting the allocator to `nullptr`. I think release functions should only relinquish the ownership of the `buf_` pointer and leave the ownership of allocator to the destructor. 

More specifically, I propose
1. `release` should transfer the ownership of `buf_` and `allocator_`  to `DetachedBuffer` and **not** reset the `allocator_` to `nullptr`. It should only disown the allocator.
2. `release_raw` should only return a pointer and not eagerly delete the `allocator_`. Nor should it set it to `nullptr`. In this case, destructor of `FlatBufferBuilder` will take care of deleting the allocator if/when needed. 
3. If clearing allocator is actually desired, there's `.clear_allocator` for that.

The reason I'm suggesting this is that grpc `MessageBuilder` has a funky member allocator and aggressively setting the allocator to `nullptr` in `release` and `release_raw` functions makes the `this` object unusable in case of `MessageBuilder`. 

In essence, the current code breaks `MessageBuilder` **is-not-a** `FlatBufferBuilder` relationship, which is very surprising. I've a test that proves that. See `builder_move_assign_after_release_test` and `builder_move_assign_after_releaseraw_test` in PR #4902. I created `GrpcLikeMessageBuilder` to test. Without the [changes](https://github.com/google/flatbuffers/pull/4902/files#diff-e644a513ebf0d4b999ed39c245f8f3db) in `vector_downward` the tests fail. With PR #4902, the `is-a` relationship is restored.

Thoughts?

gabyx commented:
Hm... I had similar doubts/thought about the semantics of deallocation with the allocator when `release_raw`. Your first point seems perfectly fine, and is just a nice feature to have (?).
The thing is, "ownership" is transferred from FlatBufferBuilder down into the `vector_downward` member. 

So you're kind of right, after
```cpp
FlatBufferBuilder builder(alloc, true,...)
...
builder.ReleaseRaw(...);
```
The allocator is destroyed which is not really a good behavior. I think (2) and (3) are both perfectly valid and should be implemented as you did :-).

sutambe commented:
I think a small correction is needed in my proposal. That's because `DetachedBuffer::Release` and `vector_downward::release` have been around since May' 17. I don't think it makes sense to change that behavior at this point. So, point 1 I mentioned earlier needs tweaks. Here're my revised thoughts.
1. If `FlatBufferBuilder` owns the allocator, `release` should transfer the ownership of `buf_` and `allocator_` to `DetachedBuffer` and set the `allocator_` to `nullptr`. This is the behavior since May'17. 
2. If `FlatBufferBuilder` owns the allocator, `release_raw` should not transfer the ownership of `buf_` and `allocator_` (it's not possible anyway) and **not** deallocate the `allocator_` nor should it be set to `nullptr`. This is a suggested change since @gabyx patch last week.
3. If `FlatBufferBuilder` does not own the allocator, both `Release` and `ReleaseRaw` should not reset the `allocator_` to `nullptr`. I think this is a safe change as something external is responsible for deleting the allocator anyway.

The latest commit in this PR incorporates the above change. It has better, larger set of tests and they work fine. However, `grpc::MessageBuilder` has some issues with this test suite (`BuilderTests`). I have a hunch of why the tests are not good enough for `grpc::MessageBuilder`.

sutambe commented:
After much more testing, I think I've something that looks reasonable to me.

In the current state, if `grpc::MessageBuilder` calls `ReleaseRaw`, `grpc::MessageBuilder` basically **silently** degenerates to  default-constructed `FlatBufferBuilder` as the `allocator_` is reset to `nullptr`. If the `MessageBuilder` object is used after `ReleaseRaw`, it allocates using the internal `DefaultAllocator`. It appears to work. But in reality, you can no longer call `MessageBuilder::ReleaseMessage()` as `grpc_slice_` is completely bypassed. This behavior breaks the expectation of the user of `MessageBuilder`. 

Specifically, the following test code should fail but does not. 
```cpp
flatbuffers::grpc::MessageBuilder b1;
for (int i = 0; i < 5; ++i) {
      auto root_offset1 = populate1(b1); // fill in m1_name and m1_color
      b1.Finish(root_offset1);
      size_t size, offset;
      const uint8_t *buf = b1.ReleaseRaw(size, offset);
      TEST_ASSERT(verify(buf, offset, m1_name, m1_color));
      // b1 now indistinguishable from FlatBufferBuilder.
}
```

With this PR, the new behavior is that calling `grpc::MessageBuilder::ReleaseRaw` has no hidden surprises and the `MessageBuilder` object is no longer usable (destructor and move-assign are allowed). No silent degeneration takes place. If someone tries to populate the `MessageBuilder` after `ReleaseRaw` is called, the program crashes with assertion failures (thanks to asserts in `SliceAllocator`). 

This PR includes a [negative test](https://github.com/google/flatbuffers/pull/4902/files#diff-e4e97de25fd24e0858e94c26bb61eb1eR35) same as above. 

The `MessageBuilder` can be brought back to life if a default-constructed `MessageBuilder` is moved into it.

aardappel commented on tests/test.cpp:
```diff
@@ -1990,25 +1954,25 @@ void LoadVerifyBinaryTest() {
-  #else
-    auto &flatbuf = flatbuf1;
-  #endif // !defined(FLATBUFFERS_CPP98_STL)
+#if !defined(FLATBUFFERS_CPP98_STL)
```
please do not reformat unrelated code, keep #preprocessor things indented. If this resulted from clang-format, please add `// clang-format: off` as required.

aardappel commented on tests/test_assert.cpp:
```diff
@@ -0,0 +1,18 @@
+
+int testing_fails = 0;
+
+void TestFail(const char *expval, const char *val, const char *exp,
```
lets keep our "testing framework" header only if possible. since there is an `assert(0)` in there, it is intended to halt on the first test failure.

sutambe replied:
It's not clear what the issue is with `assert(0)`. There are globals like `testing_fails`. I'm separating declaration and definition.  The `test.cpp` file is kind of too big.

aardappel replied:
I meant, since it is intended to halt on first failure, there's no need to count fails, and no need for a global, thus this code can be kept in the header.

sutambe replied:
I understand now. I prefer counting test failures rather than bailing out on the first error. This seems to be a norm in the gradle world (Java). But I could be persuaded otherwise.

I updated the TEST macros to include full function name as there are template instantiations. I tested without `assert(0)` and caused some tests to fail deliberately. I see output like this
```
/Users/sutambe/projects/myoss/myflatbuffers/cmake-build-debug/flattests
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<flatbuffers::FlatBufferBuilder>::builder_move_assign_after_releaseraw_test() [Builder = flatbuffers::FlatBufferBuilder]
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<TestHeapBuilder>::builder_move_assign_after_releaseraw_test() [Builder = TestHeapBuilder]
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<GrpcLikeMessageBuilder>::builder_move_assign_after_releaseraw_test() [Builder = GrpcLikeMessageBuilder]
3 FAILED TESTS

Process finished with exit code 1
```
Output on Windows is not gonna be so nice. It is still better as it tells me that 3 tests failed in all as opposed to the first one only. I'll rebase if you are ok with it or revert and drop the counter.

What say?

aardappel commented on tests/builder_test.h:
```diff
@@ -0,0 +1,260 @@
+#ifndef BUILDER_TEST_H
```
maybe call this file `test_builder.h` so it sorts with the other cpp test files.

aardappel commented on tests/builder_test.cpp:
```diff
@@ -0,0 +1,110 @@
+#include "monster_test_generated.h"
```
just stick all this code in the header for simplicity? if not, there may be other build files that need updating, e.g. `BUILD`, Android etc.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -639,18 +639,22 @@ class vector_downward {
     offset = static_cast<size_t>(cur_ - buf_);
 
+    // release_raw only relinquishes the buffer ownership.
+    // Does not deallocate or reset the allocator. Destructor will do that.
```
Ok, that seems indeed nicer than what we had before.. change the minimum possible.

aardappel commented on tests/test_assert.h:
```diff
@@ -0,0 +1,39 @@
+#ifndef TEST_ASSERT_H
```
ahh yes, nice to share this code.

sutambe replied:
Does this comment mean, you now prefer separation of test code in .h and .cpp?

aardappel replied:
No, this meant it is nice this code is shared between the main test and the gRPC test.
I still prefer things to be in headers, but I can be convinced otherwise.

aardappel commented:
@gabyx does the detach related changes look good to you?
@rw please merge once everyone is happy if I am not around.

sutambe commented:
@aardappel @rw I feel I'm pretty much done with this PR. I want to avoid making test.cpp longer that it is already. I like the separation of test code between test_builder files and previous test.cpp. There's nice cohesion between `tests/test_builder.*` and `grpc/tests/message_builder_test.cpp`

sutambe commented:
Although I mentioned earlier that I'm done modifying the PR, I made some important additions last night. These last set of changes fix what I thought was the sticking point in this PR--failing `MessageBuilder` test in `REUSABLE_AFTER_RELEASE_RAW` case. See [message_builder_test.cpp](https://github.com/google/flatbuffers/pull/4902/files#diff-e4e97de25fd24e0858e94c26bb61eb1eR51). It no longer fails. To fix it, here's what I did.

I added `MessageBuilder::ReleaseRaw` [function](https://github.com/google/flatbuffers/pull/4902/files#diff-8255ebaedac034d6a10c9b8dc1b8c272R198). The signature is 
```
 uint8_t *ReleaseRaw(size_t &size, size_t &offset, grpc_slice &slice);
```
As you can see, this function is similar in spirit to the one @gabyx added in `FlatBufferBuilder`. The similarity is that in both classes, calling this function relinquishes the ownership of the buffer pointer  to the caller. That means in `MessageBuilder`, `grpc_slice` must be returned so that the caller can free the raw pointer later by calling `grpc_slice_unref()`. After this function, `MessageBuilder` is as if it's default constructed. 

Without this new function, calling base class's `ReleaseRaw` has really surprising side-effects. The released pointer is deleted after `MessageBuilder` goes out of scope. This weird behavior caused one test to fail, which is now fixed.

sutambe commented:
@aardappel @rw What are the next steps for this PR now? 
@gabyx Do you have some cycles to take a look?

gabyx commented:
I will have a look at the detached related things, what I ve seen so far is better, yes! maybe tomorrow

gabyx commented on grpc/tests/message_builder_test.cpp:
```diff
@@ -39,121 +14,175 @@ bool release_n_verify(flatbuffers::grpc::MessageBuilder &mbb, const std::string
+      b1.Finish(root_offset1);
+      size_t size, offset;
+      grpc_slice slice;
+      const uint8_t *buf = b1.ReleaseRaw(size, offset, slice);
```
Do I understand correctly, that `slice` is owning the memory (refcount) and grpc_slice_unref destructs it...?

sutambe replied:
yes, that's my understanding.

sutambe replied:
More accurately, `grpc_slice_unref` destroys the data and refcount when the refcount hits zero. Otherwise, `slice` only refers to refcount.

gabyx commented:
@sutambe I think, you did some good work here. =)

sutambe commented:
@gabyx Thx much! Your comment made my day.

sutambe commented:
@aardappel @rw What are the next steps for this PR now?

aardappel commented:
@sutambe looks ready to be merged.. see my last comment above though.

sutambe commented:
@aardappel I did some changes in response to your comment. I left a "reply" comment last week. Reposting it here.
I understand your comment about `assert(0)`. I prefer counting test failures rather than bailing out on the first error. This seems to be a norm in the gradle world (Java). But I could be persuaded otherwise.

I updated the TEST macros to include full function name as there are template instantiations. I tested without `assert(0)` and caused some tests to fail deliberately. I see output like this
```
/Users/sutambe/projects/myoss/myflatbuffers/cmake-build-debug/flattests
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<flatbuffers::FlatBufferBuilder>::builder_move_assign_after_releaseraw_test() [Builder = flatbuffers::FlatBufferBuilder]
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<TestHeapBuilder>::builder_move_assign_after_releaseraw_test() [Builder = TestHeapBuilder]
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<GrpcLikeMessageBuilder>::builder_move_assign_after_releaseraw_test() [Builder = GrpcLikeMessageBuilder]
3 FAILED TESTS

Process finished with exit code 1
```
Output on Windows is not gonna be so nice. It is still better as it tells me that 3 tests failed in all as opposed to the first one only. I'll rebase if you are ok with it or revert and drop the counter.

What say?

aardappel commented:
I guess it is fine, we can use it as is.
Thanks for your work!

sutambe commented:
@aardappel Thx much! Closes #4897 (with improvements)

vglavnyy commented:
Probably, remove the `assert` was not good idea.
It was useful for debug.
Now need to use step-by-step debuging to see the call-stack on first fall
The test engine should have flag like `--gtest_break_on_failure`, like google-test.
This flag could be passed over `main()` arguments or as define over CMake.

aardappel commented:
@sutambe @vglavnyy I did not spot the `assert` being removed, I'd agree we want to keep that. Can you put it back?

sutambe commented:
@aardappel @vglavnyy Created PR #4951.
